### ISSUE TITLE
systemd-networkd: bridge MAC

### ISSUE THREAD

**Issue by robimarko** [systemd/systemd#19559]
Systemd-networkd has support for creating and managing network bridges as any other network interface. 

However, the resulting bridge interface will have a random MAC address assigned to it at creation based on the systemd machine-id and interface name as well as some hashing.

This will work on most network drivers as they don't check the upstream MAC address, but for some drivers, like the switchdev Marvell Prestera, that will not work.

They check the upstream interface MAC address to see whether the prefix matches one of the bridge slave interfaces as they require the prefix to be the same for the HW offloading to work. This in turn would cause the driver to error out and in turn, systemd-networkd won't be able to configure the bridge interface, thus resulting in a non-working state.

Previously, systemd would check the number of slave bridge interfaces and, if there was only one interface, the bridge would inherit the MAC address of that slave interface. Unfortunately, this was dropped in v243 release.
Original commit: https://github.com/systemd/systemd/commit/deb2cfa4c6885d448eb1f17e5ef1b139106b7e86

Currently, there is no configuration option to select the interface from which the bridge interface would inherit the MAC address. This requires the users to either manually specify the MAC address in the bridge interface configuration or set it using ifconfig, iproute, etc userspace tools.

**Comment by robimarko** [comment]
I would not say that it's a driver bug, if a bridge is used then it requires the MAC prefix to be the same, otherwise, HW offloading would not work.
HW requires that so the driver has a check for it.

It does not require the MAC to be exactly the same as one of the switch ports in the bridge, just the prefix needs to be the same.

**Comment by yegor-alexeyev** [comment]
Hi! Let me suggest another approach for this problem.
How about allowing to specify some prefix of bridge MAC address in configuration file? The rest of the mac address may be randomly generated.

**Comment by robimarko** [comment]
Hi,
unfortunately, that won't help as even the currently address can be hardcoded in the bridge configuration, however, the issue is that it's not a solution to rely on post bridge bring-up trigger to then have a service start a script that will then apply a calculated MAC and then call reconfigure on the bridge.

**Comment by yegor-alexeyev** [comment]
> > They check the upstream interface MAC address to see whether the prefix matches one of the bridge slave interfaces
> 
> Sounds like a bug in the driver, no? If the driver does such a check, it means that it requires that interface that has the same MAC as the bridge must be always present, and the bridge is not functional without it. This means that both during startup we cannot bring the bridge up, and if the interface goes away for any reason, the bridge stops working. Thus, shouldn't the driver just be fixed no to do that?

A lot of bridge setups are static and are not changing. For these that feature may be useful. Given that it may be done as an opt-in I see no harm in it.

**Comment by lperkov** [comment]
I would support this feature as well since we need it too. Could we develop a prototype and comment on the pull request?

**Comment by keszybz** [comment]
> I would not say that it's a driver bug, if a bridge is used then it requires the MAC prefix to be the same, otherwise, HW offloading would not work.

I somehow missed the part that only the prefix is checked. This sounds reasonable.

I agree, it should be fine to configure the MAC address for the bridge.
(An option to set just some prefix, and have the remaining part generated would be an option too, but that'd be more complicated, without any benefit over just configuring the whole address.)
But looking into this, we already have [NetDev] MACAddress=.
... and it seems to work just fine:

```console
$ cat /etc/systemd/network/test1.netdev
[NetDev]
Name=bridge0
Kind=bridge
MACAddress=cc:e8:50:f0:93:d8
$ sudo networkctl reload
$ ip link show dev bridge0
3: bridge0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether cc:e8:50:f0:93:d8 brd ff:ff:ff:ff:ff:ff
```

**Comment by robimarko** [comment]
Yes, you can hardcode the MAC address, but that means that you have to do it before or after bringing up the bridge.
Since there is no hook for a systemd service we currently trigger when the bridge is brought up, then service will start the script which will calculate the MAC, apply it to the bridge interface and call networkctl to reconfigure on the bridge.

This is really not efficient to do as you would have to do it for each and every bridge, why not have an opt-in property to have systemd do this before the bridge is configured?

**Comment by keszybz** [comment]
I don't understand. Initially you said the bridge MAC must match a certain prefix (the same as in the devices). So I would assume that to know that prefix, you need to know the MAC address of at least one device. And if know the full MAC address, then why not hardcode that for the whole bridge?

**Comment by keszybz** [comment]
To a large extent, this can achieve this effect by setting MACAddressPolicy=none and then letting the kernel copy the MAC address from the first enslaved interface. We could add an option like `MACAddressFrom=<iface>`, to do something similar internally in networkd with more configurability. It would be worth it if the new approach provides noticeable benefits.

So... ignoring for now the question how *easy* it would be to implement, what settings / mechanism would you like to see in systemd-networkd?

**Comment by robimarko** [comment]
Ok, so MACAddressPolicy=none would partially solve this, but it does not have persistency as there is no guarantee that the drivers will register interfaces in the same order.

`MACAddressFrom=<iface>` and potentially a method to use an offset of that MAC would be ideal.

**Comment by keszybz** [comment]
@yuwata You reworked networkd to use a request queue... How hard would it be to implement MACAddressFrom= in that scheme? If the source interface was always available, things would be easy. But since it may appear late, we would need to delay the setting of the MAC address.

**Comment by yegor-alexeyev** [comment]
Hi @yuwata. Doeth this mean that netdevs configuration feature through request queue is a precondition for the implementation of this one? May I help with it?

**Comment by robimarko** [comment]
So, after some testing, it looks that in this case, the driver always registers networking interfaces in the same order so the MAC that the bridge inherits is always the same with `MACAddressPolicy=none` so that works.

Since there is no need for implementing anything I am closing this.
Thanks for the feedback.

**Comment by yuwata** [comment]
@yegor-alexeyev If you still want your suggestion being implemented. Please open a new issue page. Thanks.

**Comment by daiaji** [comment]
I set up the link file, but my only bridge doesn't seem to be inheriting the mac address of the network card.
```
sudo cat /etc/systemd/network/51-bridge.link
[Match]
Driver=bonding bridge

[Link]
MACAddressPolicy=none
sudo cat /etc/systemd/network/50-wired.link
[Match]
MACAddress=00:d8:61:xx:xx:xx

[Link]
NamePolicy=kernel database onboard slot path
MACAddressPolicy=none
WakeOnLan=magic

2: br0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 66:df:dc:xx:xx:xx brd ff:ff:ff:ff:ff:ff
4: enp39s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel master br0 state UP group default qlen 1000
    link/ether 00:d8:61:xx:xx:xx brd ff:ff:ff:ff:ff:ff
systemctl --version                            
systemd 252 (252.2-1-manjaro)
+PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```

**Comment by yuwata** [comment]
@daiaji Could you open a new issue with debugging logs of networkd?

**Comment by daiaji** [comment]
Looks like a bug？

**Comment by yuwata** [comment]
Not sure. But I do not want to forget your report. If you comment already closed issue, then we can easily forget it.

**Comment by jalbstmeijer** [comment]
@yuwata did you ever found the reason MACAddressPolicy=none was not working for you? Seem to have the same issue.

**Comment by yuwata** [comment]
@jalbstmeijer You need to also specify MACAddress=none in .netdev file.
See https://www.freedesktop.org/software/systemd/man/systemd.netdev.html#MACAddress=

**Comment by daiaji** [comment]
Is MACAddressFrom or something like that implemented?

**Comment by jalbstmeijer** [comment]
No, you achieve that with the combination of

"MACAddressPolicy=none" in the .link file
and
"MACAddress=none" in the .netdev file

**Comment by ezickler** [comment]
I just wanted to leave this for anyone else searching for this problem as I did. 

The bridge does not take the MAC address of the first  enslaved interface as stated above but the kernel uses the lowest mac address of all the enslaved interface. As this took us a bit to figure out it will hopefully help other.

Found the answer here:`
https://superuser.com/questions/1720967/how-does-linux-determine-the-default-mac-address-of-a-bridge-device
